codegen/service/service_data.go

```diff
@@ -597,3 +699,301 @@ func requirements(m *design.MethodExpr) []*design.SecurityExpr {
+	obj := &design.Object{}
+	tgt := &design.AttributeExpr{
+		Type: &design.UserTypeExpr{
+			AttributeExpr: &design.AttributeExpr{Type: obj},
+			TypeName:      scope.GoTypeName(vres),
+		},
+	}
+	vresobj := design.AsObject(vres.Type)
+	// Select only the attributes from the given view
+	for _, n := range *view.Type.(*design.Object) {
+		if attr := vresobj.Attribute(n.Name); attr != nil {
+			// Add any specific view metadata from view attribute
+			if v, ok := n.Attribute.Metadata["view"]; ok {
+				if attr.Metadata == nil {
+					attr.Metadata = design.MetadataExpr{}
+				}
+				attr.Metadata["view"] = v
+			}
+			obj.Set(n.Name, attr)
+		}
+	}
+	code, helpers := buildConstructorCode(res, tgt, "res", "p", "", viewspkg, view.Name, scope)
+	rname := scope.GoTypeName(res)
+	name := "New" + rname + codegen.Goify(view.Name, true)
+	return &InitData{
+		Name:        name,
+		Description: fmt.Sprintf("%s projects result type %s into viewed result type %s using the %s view.", name, rname, rname, view.Name),
+		Args:        []*InitArgData{{Name: "res", Ref: scope.GoTypeRef(res)}},
+		ReturnRef:   scope.GoFullTypeRef(res, viewspkg),
+		Code:        code,
+		Helpers:     helpers,
+	}
+}
+
+func buildConstructorCode(src, tgt *design.AttributeExpr, srcvar, tgtvar, srcpkg, tgtpkg string, view string, scope *codegen.NameScope) (string, []*codegen.TransformFunctionData) {
+	// t contains only the attributes of type other than result type from tgt
+	t := design.DupAtt(tgt)
+	// trts contains only the attributes of type result type from tgt
+	trts := &design.Object{}
+	tobj := design.AsObject(t.Type)
+	for _, n := range *tobj {
+		if _, ok := n.Attribute.Type.(*design.ResultTypeExpr); ok {
+			trts.Set(n.Name, n.Attribute)
+			tobj.Delete(n.Name)
+		}
+	}
+	code, helpers, err := codegen.GoTypeTransform(src.Type, t.Type, srcvar, tgtvar, srcpkg, tgtpkg, view == "", scope)
+	if err != nil {
+		panic(err) // bug
+	}
+	data := map[string]interface{}{
+		"Source": srcvar,
+		"Target": tgtvar,
+		"Code":   code,
+		"View":   view,
+	}
+	if view != "" {
+		data["InitName"] = tgtpkg + "." + "New" + scope.GoTypeName(src)
+	}
+	attrs := make([]map[string]interface{}, 0, len(*trts))
+	for _, n := range *trts {
+		init := "New" + scope.GoTypeName(n.Attribute)
+		if view != "" {
+			v := "default"
+			if attv, ok := n.Attribute.Metadata["view"]; ok {
+				// view is explicitly set for the result type on the attribute
+				v = attv[0]
+			}
+			init += codegen.Goify(v, true)
+		}
+		varn := codegen.Goify(n.Name, true)
+		attrs = append(attrs, map[string]interface{}{
+			"SourceVar": srcvar + "." + varn,
+			"TargetVar": tgtvar + "." + varn,
+			"InitName":  init,
+		})
+	}
+	data["Attributes"] = attrs
+	var buf bytes.Buffer
+	if err = initResultTypeCodeTmpl.Execute(&buf, data); err != nil {
+		panic(err) // bug
+	}
+	return buf.String(), helpers
+}
+
+const initResultTypeCodeT = `{{- if not .View }}
```

Comment:
I don't follow the code below, looks like we initialize `{{ .TargetVar }}` but then use `{{ .Target }}`?

Reply:
`TargetVar` refers to struct fields of type result type. `Target` is the final result type/viewed result type returned.

Reply:
Modified the code to have better clarity.